include/adc.h

```diff
@@ -27,119 +27,152 @@ extern "C" {
  */
 
 /**
- * @brief ADC driver Sequence entry
- *
- * This structure defines a sequence entry used
- * to define a sample from a specific channel.
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_entry {
-	/** Clock ticks delay before sampling the ADC. */
-	s32_t sampling_delay;
-
-	/** Buffer pointer where the sample is written.*/
-	u8_t *buffer;
+struct adc_sampling {
+	/** Clock ticks delay before starting the sampling (counted from the
+	 *  start of the previous sampling in the sequence, if any). */
+	u32_t delay;
 
-	/** Length of the sampling buffer.*/
-	u32_t buffer_length;
+	/** Pointer to a buffer where the samples are to be written.
+	 *  Each sample value is sign-extended to 16 bits. */
+	s16_t *buffer;
 
-	/** Channel ID that should be sampled from the ADC */
-	u8_t channel_id;
+	/** Channels to be used for getting consecutive samples. */
+	const u8_t *channel_ids;
 
-	u8_t stride[3];
+	/** Number of samples to read in the sampling. */
+	u8_t sample_count;
 };
 
 /**
- * @brief ADC driver Sequence table
- *
- * This structure defines a list of sequence entries
- * used to execute a sequence of samplings.
+ * @brief ADC sampling definition structure.
  */
-struct adc_seq_table {
-	/* Pointer to a sequence entry array. */
-	struct adc_seq_entry *entries;
+struct adc_channel_cfg {
+	u8_t input_positive;
+	u8_t input_negative;
+	/**
+	 *  A bit field with the following parts:
+	 *  [5 :  0] - acquisition time, in microseconds
+	 *  [6]      - mode, 0: single-ended (only positive input is specified)
+	 *                   1: differential (both inputs need to be specified)
+	 *  [7 :  ?] - gain/prescaling setting
+	 *  anything else ?
+	 */
+	u16_t operation;
+};
 
-	/* Number of entries in the sequence entry array. */
-	u8_t num_entries;
-	u8_t stride[3];
+/**
+ * @brief ADC configuration structure.
+ */
+struct adc_config {
+	/** A valid pointer to an actual ADC device. */
+	struct device *dev;
+
+	/** ADC resolution.
+	 * For single-ended channels the sample values are from range:
+	 *   0 .. 2^resolution - 1,
+	 * for differential ones:
+	 *   -2^resolution .. 2^resolution - 1. */
+	u8_t resolution;
+	/** Oversampling setting.
+	 *  Each sample is averaged from 2^oversampling results. */
+	u8_t oversampling;
 };
 
 /**
  * @brief ADC driver API
  *
- * This structure holds all API function pointers.
+ * This is the mandatory API any ADC driver needs to expose.
  */
 struct adc_driver_api {
-	/** Pointer to the enable routine. */
-	void (*enable)(struct device *dev);
-
-	/** Pointer to the disable routine. */
-	void (*disable)(struct device *dev);
-
-	/** Pointer to the read routine. */
-	int (*read)(struct device *dev, struct adc_seq_table *seq_table);
+	int (*read)(const struct adc_config *config,
```

Comment:
Please add the dev pointer argument back and remove it from the adc_config struct